connector/sumconnector/sum.go

```diff
@@ -0,0 +1,159 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package sumconnector // import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/sumconnector"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+var noAttributes = [16]byte{}
+
+func roundFloat(val float64) float64 {
+	return math.Round(val*100000) / 100000
+}
+
+func newSummer[K any](metricDefs map[string]metricDef[K]) *summer[K] {
+	return &summer[K]{
+		metricDefs: metricDefs,
+		sums:       make(map[string]map[[16]byte]*attrSummer, len(metricDefs)),
+		timestamp:  time.Now(),
+	}
+}
+
+type summer[K any] struct {
+	metricDefs map[string]metricDef[K]
+	sums       map[string]map[[16]byte]*attrSummer
+	timestamp  time.Time
+}
+
+type attrSummer struct {
+	attrs pcommon.Map
+	sum   float64
+}
+
+func (c *summer[K]) update(ctx context.Context, attrs pcommon.Map, tCtx K) error {
+	var multiError error
+	for name, md := range c.metricDefs {
+		sourceAttribute := md.sourceAttr
+		sumAttrs := pcommon.NewMap()
+		var sumVal float64
+
+		// Get source attribute value
+		if sourceAttrVal, ok := attrs.Get(sourceAttribute); ok {
+			switch {
+			case sourceAttrVal.Str() != "":
+				sumVal, _ = strconv.ParseFloat(sourceAttrVal.Str(), 64)
```

Comment:
It may be worthwhile to include that the default is to convert to floats for strings in the README, but I'm not sure if there's a clear way to include it there.